### PR TITLE
impl(storage): make RestClient http transport the default

### PR DESCRIPTION
**Pull request by scotthart**
Part of the work for #9659

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)

### CODE REVIEW

File: google/cloud/storage/client_options.cc [hunk 1]
```diff
@@ -229,8 +229,12 @@ Options DefaultOptions(std::shared_ptr<oauth2::Credentials> credentials,
     o.set<ProjectIdOption>(std::move(*project_id));
   }
 
-  auto use_rest = GetEnv("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT");
-  if (use_rest.has_value()) o.set<UseRestClientOption>(true);
+  auto use_legacy_http = GetEnv("GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP");
+  if (use_legacy_http.has_value()) {
```
Comment (coryan) [inline google/cloud/storage/client_options.cc:233]: I think we want:

```cc
// The environment variable overrides values from the code
if (use_legacy_http.has_value()) {
  o.set<UseRestClientOption(false);
} else if (!o.has<UseRestClientOption>()) {
  o.set<UseRestClientOption>(true);
}
```

File: google/cloud/storage/client_options_test.cc [hunk 2]
```diff
@@ -445,20 +445,21 @@ TEST_F(ClientOptionsTest, ProjecIdtWithEnv) {
...
 
-TEST_F(ClientOptionsTest, UseRestClientWithoutEnv) {
-  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_HAVE_REST_CLIENT",
+TEST_F(ClientOptionsTest, UseRestClientByDefault) {
+  ScopedEnvironment env("GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP",
                         absl::nullopt);
   auto const options =
       internal::DefaultOptions(oauth2::CreateAnonymousCredentials(), {});
-  EXPECT_FALSE(options.has<internal::UseRestClientOption>());
+  EXPECT_TRUE(options.has<internal::UseRestClientOption>());
```
Comment (coryan) [inline google/cloud/storage/client_options_test.cc:453]: Should `UseRestClientOption` become public?  If not, why document it in the CHANGELOG?

**Review by coryan — CHANGES_REQUESTED** [review]
I think we need to change the `coverage.sh` build to test the legacy client, that can be removed with all the other code.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan and @scotthart)

___
*[`google/cloud/storage/client_options_test.cc` line 453 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9Dj8AmbwIbq39N92__-r1-453:-N9IXjI8AUfmLhLIDAdv:bq9ghsw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8805e4936fe832526c5c83e1b47bc1de81e39a9/google/cloud/storage/client_options_test.cc#L453)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Should `UseRestClientOption` become public?  If not, why document it in the CHANGELOG?
</blockquote></details>

For some reason I thought it was public. Removed mention of it from the CHANGELOG.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)* status: 0 of 6 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/storage/client_options.cc` line 233 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9Dj8AmbwIbq39N92_Z-r1-233:-N9IXqFS2-XsPi4EgDrz:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8805e4936fe832526c5c83e1b47bc1de81e39a9/google/cloud/storage/client_options.cc#L233)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think we want:

```cc
// The environment variable overrides values from the code
if (use_legacy_http.has_value()) {
  o.set<UseRestClientOption(false);
} else if (!o.has<UseRestClientOption>()) {
  o.set<UseRestClientOption>(true);
}
```

</blockquote></details>

Done.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 3 of 5 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`CHANGELOG.md` line 131 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9OfksRAFsnr7lTKjxa:-N9OfksS7hcTGMzYpAAe:b-kvbyxk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7ae875abc3dd2024730169cece99caece330486c/CHANGELOG.md#L131)):*
> ```Markdown
> ### [Storage](/google/cloud/storage/README.md)
> 
> * The internal http implementation has changed and provides comparable download
> ```

Sorry for the late comments.  ~The passive voice should be avoided~. I think we should use the active voice:

```md
* We have a new implementation of the HTTP transfer. This new
   implementation provides comparable download performance,
   and improves some uploads.  See [#9659] for details.
   We have rigorously tested this new implementation. In the
   unlikely event that this new implementation breaks your application,
   we included an environment variable to revert to the legacy implementation.
   Set `GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP` to any
   value to use the legacy implementation.
   We are planning to remove the legacy code and implementation by 2022-12

[#9659]: https://github.com/googleapis/google-cloud-cpp/issues/9659
```

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)* status: 5 of 6 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`CHANGELOG.md` line 131 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9OfksRAFsnr7lTKjxa:-N9WrmsdC4DhUk-L_BE9:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/7ae875abc3dd2024730169cece99caece330486c/CHANGELOG.md#L131)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Sorry for the late comments.  ~The passive voice should be avoided~. I think we should use the active voice:

```md
* We have a new implementation of the HTTP transfer. This new
   implementation provides comparable download performance,
   and improves some uploads.  See [#9659] for details.
   We have rigorously tested this new implementation. In the
   unlikely event that this new implementation breaks your application,
   we included an environment variable to revert to the legacy implementation.
   Set `GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP` to any
   value to use the legacy implementation.
   We are planning to remove the legacy code and implementation by 2022-12

[#9659]: https://github.com/googleapis/google-cloud-cpp/issues/9659
```


</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680)* status: all files reviewed, 2 unresolved discussions (waiting on @scotthart)

___
*[`CHANGELOG.md` line 131 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9WtpezB3WLkW6tTk3U:-N9WtpezB3WLkW6tTk3V:bikbhof) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/aff22ec09d974378b404309ffd9e8c188f9fc509/CHANGELOG.md#L131)):*
> ```Markdown
> ### [Storage](/google/cloud/storage/README.md)
> 
> * We have a new implementation of the HTTP transfer. This new implementation
> ```

My wording was not great I think.  Maybe "We have a new implement of HTTP transfers"? or "HTTP requests"?
___
*[`CHANGELOG.md` line 137 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9680#-N9WtzPQ0r3rGgxZxu9y:-N9WtzPQ0r3rGgxZxu9z:b-l6ko1t) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/aff22ec09d974378b404309ffd9e8c188f9fc509/CHANGELOG.md#L137)):*
> ```Markdown
>   included an environment variable to revert to the legacy implementation. Set
>   `GOOGLE_CLOUD_CPP_STORAGE_USE_LEGACY_HTTP` to any value to use the legacy
>   implementation. We are planning to remove the legacy code and implementation
> ```

s/implementation/this environment variable/ ?

**Pull request merged**